sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/BeamSql.java

```diff
@@ -47,22 +36,23 @@
 Pipeline p = Pipeline.create(options);
 
 //create table from TextIO;
-PCollection<BeamSqlRow> inputTableA = p.apply(TextIO.read().from("/my/input/patha"))
-    .apply(...);
-PCollection<BeamSqlRow> inputTableB = p.apply(TextIO.read().from("/my/input/pathb"))
-    .apply(...);
+PCollection<Row> inputTableA = p.apply(TextIO.read().from("/my/input/patha")).apply(...);
+PCollection<Row> inputTableB = p.apply(TextIO.read().from("/my/input/pathb")).apply(...);
 
 //run a simple query, and register the output as a table in BeamSql;
 String sql1 = "select MY_FUNC(c1), c2 from PCOLLECTION";
-PCollection<BeamSqlRow> outputTableA = inputTableA.apply(
-    BeamSql.query(sql1)
-    .withUdf("MY_FUNC", MY_FUNC.class, "FUNC"));
+PCollection<Row> outputTableA = inputTableA.apply(
+    BeamSql
+        .query(sql1)
+        .registerUdf("MY_FUNC", MY_FUNC.class, "FUNC")
+        .toPTransform());
```

Comment:
`toPTransform()` seems redundant from the point of view of the user. Why can't it be an immutable builder?

Reply:
I wanted to make conversion to `PTransform` explicit so that it converts mutable builder to immutable transforms, and it's clear to the user what's happening. The builder is currently mutable because it holds and mutates an instance of `BeamSqlEnv`. Refactoring `BeamSqlEnv` is too involved to include it in this PR. And gathering the required data to build `BeamSqlEnv` instance later seemed overly complex for this case. Will try to implement this approach.

Reply:
One way to get an immutable builder for a mutable object is to keep a List<Function<YourMutableObject, Void>> where each function is a mutator, applied in order when it's time to construct the mutable object (in this case probably in `expand`) :)

Reply:
I had a question on this as well - in the PTransform style guide, having an explicit builder with a `build()` step seems consistent with the consideration of when to have a separate configuration object.